src/src/com/microsoft/aad/adal/ApplicationReceiver.java

```diff
@@ -131,45 +160,102 @@ public static void clearUserName(Context ctx) {
+        {
+            @Override
+            public Object call() 
+            {
+                Logger.v(TAG + methodName, "Running task in thread:" + android.os.Process.myTid() + ", trying to get intent for "
+                        + "broker activity.");
+                BrokerProxy brokerProxy = new BrokerProxy(context);
+                final Intent resumeIntent = brokerProxy.getIntentForBrokerActivity(pendingRequest);
+                resumeIntent.setAction(Intent.ACTION_PICK);
+                
+                Logger.v(TAG + methodName, "Setting flag for broker resume request for calling package " + context.getPackageName());
+                resumeIntent.putExtra(AuthenticationConstants.Broker.BROKER_REQUEST_RESUME,
                 AuthenticationConstants.Broker.BROKER_REQUEST_RESUME);
-        intent.setPackage(AuthenticationSettings.INSTANCE.getBrokerPackageName());
-        intent.setClassName(AuthenticationSettings.INSTANCE.getBrokerPackageName(),
-                AuthenticationSettings.INSTANCE.getBrokerPackageName()
-                        + ".ui.AccountChooserActivity");
-
-        PackageManager packageManager = ctx.getPackageManager();
+                resumeIntent.putExtra(AuthenticationConstants.Broker.CALLER_INFO_PACKAGE, context.getPackageName());
 
-        // Get activities that can handle the intent
-        List<ResolveInfo> activities = packageManager.queryIntentActivities(intent, 0);
+                final String brokerProtocolVersion = resumeIntent.getStringExtra(AuthenticationConstants.Broker.BROKER_VERSION);
+                if (StringExtensions.IsNullOrBlank(brokerProtocolVersion))
+                {
+                    Logger.v(TAG + methodName, "Broker request resume is not supported in the older version of broker.");
+                    return null;
+                }
+                
+                PackageManager packageManager = context.getPackageManager();
+                // Get activities that can handle the intent
+                List<ResolveInfo> activities = packageManager.queryIntentActivities(resumeIntent, 0);
 
-        // Check if 1 or more were returned
-        boolean isIntentSafe = activities.size() > 0;
+                // Check if 1 or more were returned
+                boolean isIntentSafe = activities.size() > 0;
 
-        if (isIntentSafe) {
-            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
-            ctx.startActivity(intent);
+                if (isIntentSafe) 
+                {
+                    if (!isRunningInForeground(context, installedPackageName))
+                    {
+                        Logger.v(TAG + methodName, "Broker " + installedPackageName + " is not running in foreground yet. "
+                            + "Thread will be sleeping for 2 seconds to wait for broker package going foreground.");
+                        try 
+                        {
+                            Thread.sleep(2000);
```

Comment:
I have a older, slower device. We should test this on it. In any case, I think we should do retries with shorter intervals.